### PR TITLE
cpu/samd21: added peripheral PWM driver

### PR DESCRIPTION
**Pull request by haukepetersen**
This PR is based on the work of @PeterKietzmann, thanks for your prior work! Great job!

I started a new PR which supersedes #2473 as it is easier to handle for me and so I can offload some work from @PeterKietzmann...

@kaspar030: have a look at the parameter notation in the `periph_conf.h`, what do you think about this? When we would now allow the CPU to define the `pwm_t` type to `typedef pwm_conf_t pwm_t`, we could even simplify the code a little further...

### CODE REVIEW

File: boards/samr21-xpro/include/periph_conf.h [hunk 1]
```diff
@@ -77,6 +81,58 @@ extern "C" {
...
+    pwm_conf_chan_t chan[2];    /**< channel configuration */
+} pwm_conf_t;
+
+/* PWM device configuration */
+#if PWM_NUMOF
+static const pwm_conf_t pwm_config[] = {
+#if PWM_0_EN
+    {TCC1, {
+        {(PortGroup *)0x41004400, 6, 4, 0}, /* port 0, pin 6, af E, chan 0 */
+        {(PortGroup *)0x41004400, 7, 4, 1}  /* port 0, pin 7, af E, chan 1 */
```
Comment (PeterKietzmann) [inline boards/samr21-xpro/include/periph_conf.h]: This is maybe a stupid question but where do I find these pins?

Comment (PeterKietzmann) [inline boards/samr21-xpro/include/periph_conf.h]: Ok really stupid. Forget about it :-)

Comment (PeterKietzmann) [inline boards/samr21-xpro/include/periph_conf.h]: Please don't forget that we also need some pins for the ADC #2063. (Currently I'm not thinking about the "new-peripheral" design)

File: boards/samr21-xpro/include/periph_conf.h [hunk 2]
```diff
@@ -77,6 +81,58 @@ extern "C" {
...
+    }},
+#endif
+#if PWM_1_EN
+    {TCC0, {
+        {(PortGroup *)0x41004400, 18, 5, 2},
+        {(PortGroup *)0x41004400, 19, 5, 3}
```
Comment (PeterKietzmann) [inline boards/samr21-xpro/include/periph_conf.h]: Currently there is `GPIO_3` on that pin.

File: cpu/samd21/periph/pwm.c [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+#include "board.h"
+#include "periph/pwm.h"
+#include "periph_conf.h"
+
+/* ignore file in case no PWM devices are defined */
+#if PWM_NUMOF
+
+static inline int _num(pwm_t dev)
+{
+    return ((int)(pwm_config[dev].dev) & 0xc00) >> 10;
```
Comment (PeterKietzmann) [inline cpu/samd21/periph/pwm.c]: I still didn't internalise the new architecture. Can you explain this function shortly? Maybe a small comment would also be helpful for others

**Comment by PeterKietzmann** [comment]
@haukepetersen I successfully tested the pwm on all channels with different cycles and left/right aligned mode. With my cheap logic analyzer I saw some __short__ dropouts  (i.e. not longer than one cycle) of the signal. These appeared really sporadic so maybe it's caused by my setup. I don't know if we should merge this PR now, as #3095 seems to be ongoing and this approach would introduce a somehow different driver-design, comparing to others.

File: cpu/samd21/periph/pwm.c (hunk 3 above)
Comment (haukepetersen) [inline cpu/samd21/periph/pwm.c]: simple: if you compare the bases address of all available TCC devices and look for their differences, youll see that they differ only in bits 10 and 11. So masking these bits and shifting them to the LSB positions, youll have upcounting integers 0, 1, 2 for the three TCC devices.

**Comment by haukepetersen** [comment]
Actually this PR is completely independent from #3095. Also it is not a different design, just a mere more efficient implementation (with more room for improvement when changing the interface very slightly). The main idea about the driver model was always, that each CPU is free to implement the peripheral drivers as it sees fit (aka the most efficient way possible) while keeping to the API. So this is what this PWM implementation makes use of.

Regarding the dropouts: I was not able to see them at my (cheap) logic analyzer, so my guess would be a measurement problem. I was also running the PWM for a while to control a motor and a server which both worked without trouble, thus strengthen that idea. Further there is nothing in the driver explaining this behavior, right?!

File: boards/samr21-xpro/include/periph_conf.h (hunk 1 above)
Comment (haukepetersen) [inline boards/samr21-xpro/include/periph_conf.h]: I know. I actually needed the 4 PWM pins to be on the same connector for my car project, but I can remap them for this PR and create my project-specific mapping on the side.

File: boards/samr21-xpro/include/periph_conf.h (hunk 2 above)
Comment (haukepetersen) [inline boards/samr21-xpro/include/periph_conf.h]: true. Best would be to find a generic pin mapping that fits all. I started a pin mapping document we can use to discuss:
https://docs.google.com/spreadsheets/d/1350pEm6-6vR7L_JGR5rHhaNbMrZgzkCgA-OS0akPH6U/edit#gid=1766063117

Comment (PeterKietzmann) [inline boards/samr21-xpro/include/periph_conf.h]: No access rights

File: boards/samr21-xpro/include/periph_conf.h (hunk 1 above)
Comment (PeterKietzmann) [inline boards/samr21-xpro/include/periph_conf.h]: Now I remember :-), thanks! Would you still add a small documenting comment to the code? No need to answer, just do it or not ;-)

File: cpu/samd21/periph/pwm.c (hunk 3 above)
Comment (PeterKietzmann) [inline cpu/samd21/periph/pwm.c]: Now I remember :-), thanks! Would you still add a small documenting comment to the code? No need to answer, just do it or not ;-)

**Comment by PeterKietzmann** [comment]
Well, basically you are right so I won't block this improvement. Also nobody else cares about this "different" approach which is a good sign. I agree that there is nothing in the driver explaining this behaviour but as you know, sometimes things happen that you don't have expected! Anyway, I'm convinced that these dropouts were caused by a measurement problem

**Comment by haukepetersen** [comment]
Alright, so shoud I squash? I would suggest then to merge this and take the pinout discussion somewhere else.

**Comment by PeterKietzmann** [comment]
Öhm, can we at least solve to conflict of `GPIO_3` and `PWM_1_CH_2`? E.g. by moving one of them to another pin. As far as I've seen it, `PA14` and `PB03` are still unused. Just want to avoid potential conflicts with this PR. We can start working on an optimized mapping in another PR and in the list you sent.

**Comment by haukepetersen** [comment]
of course (just forgot to do so before my last comment). Moved GPIO_3 to PB03.

**Comment by PeterKietzmann** [comment]
ACK when squashed and happy Travis. @kaspar030 do you want to comment on the parameter notation in the "periph_conf.h`?

**Comment by kaspar030** [comment]
> @kaspar030 do you want to comment on the parameter notation in the "periph_conf.h`?

Looks good, but now that you're asking, I would take the values out of the comments. I'll PR to show.

**Comment by haukepetersen** [comment]
> Parameter notation looks good to me. port + pin maybe can be one variable once hauke's new gpio implementation is merged.

Yes, that is the plan.

**Comment by haukepetersen** [comment]
added @kaspar030's fix and squashed

**Comment by haukepetersen** [comment]
And regarding the configuration in the periph_conf.h: Those structs that are defined there should of course also be moved into the cpus `periph_cpu.h` header once the gpio-remodelling PR is merged...

**Comment by PeterKietzmann** [comment]
Then let's ACK and go!

**Comment by iseitani** [comment]
Hello,
I have a question, you acquired the expected frequencies when you used a logical analyzer? 
I just cloned the master rep and when I measured the frequency using an oscilloscope instead of the expected, as its mentioned in the readme file 1kHz, I received a frequency of 47-76kHz with period 19-24 μs.
Any idea what went wrong? (I have already validated that the oscilloscope works correctly)

**Pull request merged**